### DISCUSSION TITLE
How to control breadth-first vs depth-first in dynamic graph execution

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#8871] [Q&A] [answered]
Hi Dagster team,
I'm running a pipeline in which a subgraph is called dynamically. The subgraph, which is ran for 4 dynamic outputs in the following screenshots, is running breadth first, i.e. for each run of the subgraph the first op is finished, then the second, etc. I'm looking to control the concurrency of these subgraphs.
• Is there a way to run the different subgraphs in a depth first manner (order of red boxes in the first screenshot)?
• Is there a way to control the concurrency of the subgraphs that are running breadth first, i.e. running the first two subgraphs concurrently and once those are finished, run the second batch (second screenshot)?
Thanks ! :slightly_smiling_face:

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1657706549459619?thread_ts=1657706549.459619&cid=C01U954MEER).*

**Comment by yuhan** [comment, marked as answer]
You can set priority per op level, where ops with higher priorities would run first:
```python
@op(tags={"dagster/priority": "0"})
def later():
    pass

@op(tags={"dagster/priority": "1"})
def first():
    pass
```

So with dynamic orchestration, you can do something like:

```python
@op(out=DynamicOut())
def generate_things():
    for i in range(3):
        yield DynamicOutput(i, mapping_key=str(i))


@op(tags={"dagster/priority": 1})
def downstream_one(i):
    print("1", i)
    return i


@op(tags={"dagster/priority": 2})
def downstream_two(i):
    print("2", i)
    return i


@job
def my_job():
    generate_things().map(lambda x: downstream_two(downstream_one(x)))
```

Then, you'll get a depth-first execution:
<img src="https://user-images.githubusercontent.com/4531914/178846800-fba2c64a-0dd9-4b44-9fc2-08d1b50a56f7.png" width=400/>
